app/src/main/java/com/epfl/drawyourpath/userProfile/cache/UserModelCached.kt

```diff
@@ -0,0 +1,428 @@
+            if (it) {
+                cache.updatePathsGoal(currentUserID!!, nbOfPathsGoal)
+            }
+            it
+        }
+    }
+
+    /**
+     * This function will set the photo as the profile photo of the user
+     * @param photo that we want to set
+     * @return a completable future that indicate if the photo was correctly stored
+     */
+    fun setProfilePhoto(photo: Bitmap): CompletableFuture<Boolean> {
+        checkCurrentUser()
+        return CompletableFuture.supplyAsync {
+            database.setProfilePhoto(photo).join()
+        }.thenApplyAsync {
+            if (it) {
+                cache.updatePhoto(currentUserID!!, UserEntity.fromBitmapToByteArray(photo))
+            }
+            it
+        }
+    }
+
+    private fun checkCurrentUser(checkIfNull: Boolean = true) {
+        if (database is MockDataBase) {
+            return //already a test
+        }
+        if ((checkIfNull && currentUserID == null) || currentUserID == MockAuth.MOCK_USER.getUid()) {
+            // if current user null then it is a test
+            setDatabase(MockDataBase())
+            Toast.makeText(getApplication(), R.string.toast_test_error_message, Toast.LENGTH_LONG).show()
+        }
+    }
+
+}
+
+/**
+ * helper function to create a UserModel from a UserData
+ * @param data the userData
+ * @param database the database
+ * @return the resulting UserModel
+ */
+private fun fromUserDataToUserModel(data: UserEntity, database: Database): UserModel {
+    return UserModel(
+        data.userId,
+        data.emailAddress,
+        data.username,
+        data.firstname,
+        data.surname,
+        data.getDateOfBirthAsLocalDate(),
+        data.distanceGoal,
+        data.activityTimeGoal,
+        data.nbOfPathsGoal,
+        data.getProfilePhotoAsBitmap(),
+        database
+    )
+}
+
+/**
+ * helper function to create a UserData from a UserModel
+ * @param userModel the userModel
+ * @return the resulting UserData
+ */
+private fun fromUserModelToUserData(userModel: UserModel): UserEntity {
+    return UserEntity(
+        userModel.getUserId(),
+        userModel.getUsername(),
+        userModel.getEmailAddress(),
+        userModel.getFirstname(),
+        userModel.getSurname(),
+        UserEntity.fromLocalDateToLong(userModel.getDateOfBirth()),
+        userModel.getDistanceGoal(),
+        userModel.getActivityTime(),
+        userModel.getNumberOfPathsGoal(),
+        UserEntity.fromBitmapToByteArray(userModel.getProfilePhoto())
+    )
+}
+
+/**
+ * Helper function to check if the name format of a given variableName is correct and throw directly an error if it is incorrect
+ * @param name to be check
+ * @param variableName to be checked
+ * @throw an error if the format is not correct
+ */
+private fun checkNameFormat(name: String, variableName: String) {
+    if (name.find { !it.isLetter() && it != '-' } != null || name.isEmpty()) {
+        throw java.lang.Error("Incorrect $variableName")
+    }
+}

```

Comment:
All this functions are already present in the UserModel class(or similar) try to used a UserModel as parameter in the creation of the user cache to evict code duplication.